tests/test_modemmanager.py

```diff
@@ -0,0 +1,159 @@
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.start_system_bus()
+        cls.dbus_con = cls.get_dbus(True)
+
+    def setUp(self):
+        super().setUp()
+        (self.p_mock, self.p_obj) = self.spawn_server_template("modemmanager", {}, stdout=subprocess.PIPE)
+
+    def tearDown(self):
+        if self.p_mock:
+            self.p_mock.stdout.close()
+            self.p_mock.terminate()
+            self.p_mock.wait()
+
+        super().tearDown()
+
+    def get_property(self, name):
+        return self.p_obj.Get(self.dbus_interface, name, dbus_interface=dbus.PROPERTIES_IFACE)
+
+
+@unittest.skipUnless(have_mmcli, "mmcli utility not available")
+class TestModemManagerMmcliBase(TestModemManagerBase):
+    """Base ModemManager interface tests using mmcli"""
+
+    p_mmcli = None
+
+    def start_mmcli(self, args):
+        self.assertIsNone(self.p_mmcli)
+        # pylint: disable=consider-using-with
+        self.p_mmcli = subprocess.Popen(["mmcli"] + args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        flags = fcntl.fcntl(self.p_mmcli.stdout, fcntl.F_GETFL)
+        fcntl.fcntl(self.p_mmcli.stdout, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
+    def stop_mmcli(self):
+        self.assertIsNotNone(self.p_mmcli)
+        self.assertEmptyOutput()
+        self.p_mmcli.stdout.close()
+        self.p_mmcli.terminate()
+        self.p_mmcli.wait()
+
+    def tearDown(self):
+        if self.p_mmcli:
+            self.stop_mmcli()
+        super().tearDown()
+
+    def assertOutputContains(self, expected_lines, max_wait=2000):
+        self.assertIsNotNone(self.p_mmcli)
+        start_time = int(time.time() * 1000)
+        for line in expected_lines:
+            output = None
+            while True:
+                output = self.p_mmcli.stdout.readline()
+                if output:
+                    break
+                self.assertLessEqual(int(time.time() * 1000) - start_time, max_wait, msg="Timeout exceeded")
+            self.assertEqual(output.decode("utf-8"), f"{line}\n")
+
+    def assertOutputEquals(self, expected_lines, max_wait=2000):
+        self.assertOutputContains(expected_lines, max_wait)
+        self.assertEmptyOutput()
+
+    def assertEmptyOutput(self, max_wait=100):
+        start_time = int(time.time() * 1000)
+        while int(time.time() * 1000) - start_time < max_wait:
+            self.assertFalse(self.p_mmcli.stdout.readline(), msg="Unexpected output")
+
+
+class TestModemManagerModemMmcli(TestModemManagerMmcliBase):
+    """main ModemManager interface tests using mmcli"""
+
+    def test_no_modems(self):
+        self.start_mmcli(["-m", "any"])
+
+        self.assertOutputContains(["error: couldn't find modem"])
+
+    def test_modem(self):
+        (modem_path, sim_path) = self.p_obj.AddSimpleModem()
+        self.start_mmcli(["-m", "any"])
+        self.assertOutputEquals(
+            [
+                "  -----------------------------",
```

Comment:
How stable is this CLI? Feels a bit amibitious to pin it down to every piece of ASCII art and formatting. It feels a bit more robust to check individual properties with `self.assertIn()` or possibly `assertRegex()`.

But not a veto, this can be relaxed if/when it breaks.